### PR TITLE
Optionally Avoid recomputing features

### PR DESCRIPTION
**Pull request by justusschock**
## What does this PR do?
Optionally avoids recomputing the features.

## Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to update the docs?
- [ ] Did you write any new necessary tests?

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: torchmetrics/metric.py [hunk 1]
```diff
@@ -394,7 +394,7 @@ def compute(self) -> Any:
         """Override this method to compute the final metric value from state variables synchronized across the
         distributed backend."""
 
-    def reset(self) -> None:
+    def reset(self, exclude_states: Optional[Sequence[str]] = None) -> None:
         """This method automatically resets the metric state variables to their default value."""
```
Comment (SkafteNicki) [inline torchmetrics/metric.py:398]:
```suggestion
        """This method automatically resets the metric state variables to their default value.
        Args:
            exclude_stetes: sequence of strings indicating metric states that should not be reset.
        """
```

File: torchmetrics/image/fid.py [hunk 2]
```diff
@@ -248,6 +252,13 @@ def __init__(
         self.add_state("real_features", [], dist_reduce_fx=None)
         self.add_state("fake_features", [], dist_reduce_fx=None)
 
+        if reset_real_features:
+            exclude_states = ()
+        else:
+            exclude_states = ("real_features",)
```
Comment (SkafteNicki) [inline torchmetrics/image/fid.py:258]:
```suggestion
        exclude_states = () if reset_real_features else ("real_features", )
```

File: torchmetrics/image/kid.py [hunk 3]
```diff
@@ -235,6 +239,13 @@ def __init__(
         self.add_state("real_features", [], dist_reduce_fx=None)
         self.add_state("fake_features", [], dist_reduce_fx=None)
 
+        if reset_real_features:
+            exclude_states = ()
+        else:
+            exclude_states = ("real_features",)
```
Comment (SkafteNicki) [inline torchmetrics/image/kid.py:245]:
```suggestion
        exclude_states = () if reset_real_features else ("real_features", )
```

**Review by SkafteNicki — COMMENTED** [review]
`exclude_states` should also be added to:
* `MetricCollection`
* `MinMaxMetric`
* `MultioutputWrapper`
* `MetricTracker`

for consistency

**Comment by Borda** [comment]
is it related to #709? :rabbit:

**Comment by justusschock** [comment]
@Borda waiting for some user feedback on slack :) 

I think not directly related to #709

**Comment by Borda** [comment]
> waiting for some user feedback on slack :)

what kind of feedback, could you please remind me of the thread... :rabbit:

**Comment by ebartrum** [comment]
I believe this currently breaks loading from checkpoint. See here

**Comment by ebartrum** [comment]
[I believe this currently breaks loading from checkpoint](https://pytorch-lightning.slack.com/archives/CRBLFHY79/p1644866367931579)

File: torchmetrics/image/fid.py [hunk 4]
```diff
@@ -288,3 +296,6 @@ def compute(self) -> Tensor:
 
         # compute fid
         return _compute_fid(mean1, cov1, mean2, cov2).to(orig_dtype)
+
+    def reset(self, exclude_states: Optional[Sequence[str]] = None) -> None:
```
Comment (justusschock) [inline torchmetrics/image/fid.py:300]:
```suggestion
    def reset(self, exclude_states: Optional[Sequence[str]] = None) -> None:
        exclude_states = exclude_states or ()
```

**Review by justusschock — COMMENTED** [review]
To avoid bug when passing None

**Comment by SkafteNicki** [comment]
@justusschock instead of making this PR have influence on the base class, what about re-implementing just `reset` in the relevant metrics:
```python
def reset(self):
  if not self.reset_real_features:
    # remove temporarily to avoid resetting
    value = self._defaults.pop('real_features')
    super().reset()
    self._defaults['real_features'] = value
  else:
    super().reset()
```
this feature seems to only be relevant to `Inception`, `FID` and `KID`

**Comment by justusschock** [comment]
@SkafteNicki works for me. However, afaik there are some similar metrics in NLP which might be added some day. This is why I aimed for a more general approach here. I'm fine with both, just let me know your preference :)

**Comment by SkafteNicki** [comment]
@justusschock I would go with the more local solution for the affected metrics for now. If we then in the future have other metrics that need a similar feature then I am fine with revisiting this :]

**Review by SkafteNicki — APPROVED** [review]

**Review by Borda — APPROVED** [review]

**Comment by justusschock** [comment]
@SkafteNicki Thanks for updating. Seems you were faster :) LGTM :)

**Pull request merged**